### ISSUE TITLE
azure_rm_virtualmachine fails in AzureChinaCloud

### ISSUE THREAD

**Issue by dbanttari** [ansible/ansible_collections_azure#1048]
##### SUMMARY
something called by azure_rm_virtualmachine does not obey AZURE_CLOUD_ENVIRONMENT and attempts to connect to AzureCloud when operating on AzureChinaCloud hosts

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
azure_rm_virtualmachine

##### ANSIBLE VERSION

```paste below
ansible-playbook [core 2.12.5.post0]
  config file = /ansible/ansible.cfg
  configured module search path = ['/home/runner/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.8/site-packages/ansible
  ansible collection location = /home/runner/.ansible/collections:/ansible/collections
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.8.13 (default, Jun 24 2022, 15:27:57) [GCC 8.5.0 20210514 (Red Hat 8.5.0-13)]
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION

```paste below
- name: azure.azcollection
  version: '==1.14.0'
```

##### CONFIGURATION

```paste below
ANY_ERRORS_FATAL(/ansible/ansible.cfg) = False
COLLECTIONS_PATHS(/ansible/ansible.cfg) = ['/home/runner/.ansible/collections', '/ansible/collections']
DEFAULT_FORKS(/ansible/ansible.cfg) = 50
DEFAULT_GATHER_SUBSET(/ansible/ansible.cfg) = ['all']
DEFAULT_GATHER_TIMEOUT(/ansible/ansible.cfg) = 20
DEFAULT_HASH_BEHAVIOUR(/ansible/ansible.cfg) = merge
DEFAULT_TIMEOUT(/ansible/ansible.cfg) = 20
DEFAULT_TRANSPORT(/ansible/ansible.cfg) = smart
HOST_KEY_CHECKING(/ansible/ansible.cfg) = False
INVENTORY_ENABLED(/ansible/ansible.cfg) = ['vue.azure.azure_rm', 'script', 'yaml', 'ini']
PERSISTENT_COMMAND_TIMEOUT(/ansible/ansible.cfg) = 20
```

##### OS / ENVIRONMENT
running on a recent awx-ee image that has had additional packages/modules installed

##### STEPS TO REPRODUCE
example one: setting vm tags
```yaml
- name: Set patch_version tags
  azure.azcollection.azure_rm_virtualmachine:
    resource_group: "{{ resource_group }}"
    name: "{{ name }}"
    tags:
      patch_version: "{{ ansible_date_time.iso8601 }}"
    zones: "{{ availability_zone }}"
  delegate_to: localhost
  when:
    - "'patch_version' in tags"
    - tags['patch_version'] == 'none'
```
example two: rebooting a vm via azure_rm_virtualmachine:
```
  - name: Tell Azure to reboot this VM (so UAC setting change takes effect)
    azure.azcollection.azure_rm_virtualmachine:
      resource_group: "{{ resource_group }}"
      name: "{{ name }}"
      restarted: yes
      zones: "{{ availability_zone }}"
    delegate_to: localhost
    connection: local
```

##### EXPECTED RESULTS
VM tags set, or VM rebooted, no fatal errors

##### ACTUAL RESULTS
```paste below
TASK [windows : Set patch_version tags] ***************************************************************************************************************
task path: /ansible/roles/windows/tasks/patching_apply.yml:32
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<localhost> EXEC /bin/sh -c 'echo ~root && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344 `" && echo ansible-tmp-1671211081.1372511-135-112305740467344="` echo /root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344 `" ) && sleep 0'
Using module file /home/runner/.ansible/collections/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py
<localhost> PUT /home/runner/.ansible/tmp/ansible-local-64whuibyec/tmpa3z1tq85 TO /root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/AnsiballZ_azure_rm_virtualmachine.py
<localhost> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/ /root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/AnsiballZ_azure_rm_virtualmachine.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python3 /root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/AnsiballZ_azure_rm_virtualmachine.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/ > /dev/null 2>&1 && sleep 0'
fatal: [************-vm0 -> localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "ClientSecretCredential.get_token failed: Authentication failed: AADSTS500011: The resource principal named https://management.azure.com was not found in the tenant named PVUECN. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.\r\nTrace ID: ****\r\nCorrelation ID: ****\r\nTimestamp: 2022-12-16 17:18:11Z\nTraceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/AnsiballZ_azure_rm_virtualmachine.py\", line 107, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/AnsiballZ_azure_rm_virtualmachine.py\", line 99, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1671211081.1372511-135-112305740467344/AnsiballZ_azure_rm_virtualmachine.py\", line 47, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.azure.azcollection.plugins.modules.azure_rm_virtualmachine', init_globals=dict(_module_fqn='ansible_collections.azure.azcollection.plugins.modules.azure_rm_virtualmachine', _modlib_path=modlib_path),\n  File \"/usr/lib64/python3.8/runpy.py\", line 207, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.8/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib64/python3.8/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_4tp4ffhc/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 2344, in <module>\n  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_4tp4ffhc/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 2340, in main\n  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_4tp4ffhc/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 963, in __init__\n  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_4tp4ffhc/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/module_utils/azure_rm_common.py\", line 469, in __init__\n  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_4tp4ffhc/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 1114, in exec_module\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/tracing/decorator.py\", line 78, in wrapper_use_tracer\n    return func(*args, **kwargs)\n  File \"/usr/local/lib/python3.8/site-packages/azure/mgmt/compute/v2021_04_01/operations/_virtual_machines_operations.py\", line 1502, in get\n    pipeline_response = self._client._pipeline.run(request, stream=False, **kwargs)\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 211, in run\n    return first_node.send(pipeline_request)  # type: ignore\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 71, in send\n    response = self.next.send(request)\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 71, in send\n    response = self.next.send(request)\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 71, in send\n    response = self.next.send(request)\n  [Previous line repeated 2 more times]\n  File \"/usr/local/lib/python3.8/site-packages/azure/mgmt/core/policies/_base.py\", line 47, in send\n    response = self.next.send(request)\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_redirect.py\", line 158, in send\n    response = self.next.send(request)\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_retry.py\", line 446, in send\n    response = self.next.send(request)\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_authentication.py\", line 116, in send\n    self.on_request(request)\n  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_authentication.py\", line 93, in on_request\n    self._token = self._credential.get_token(*self._scopes)\n  File \"/usr/local/lib/python3.8/site-packages/azure/identity/_internal/get_token_mixin.py\", line 76, in get_token\n    token = self._request_token(*scopes, **kwargs)\n  File \"/usr/local/lib/python3.8/site-packages/azure/identity/_internal/decorators.py\", line 56, in wrapper\n    return fn(*args, **kwargs)\n  File \"/usr/local/lib/python3.8/site-packages/azure/identity/_internal/client_credential_base.py\", line 40, in _request_token\n    raise ClientAuthenticationError(message=message)\nazure.core.exceptions.ClientAuthenticationError: Authentication failed: AADSTS500011: The resource principal named https://management.azure.com was not found in the tenant named PVUECN. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.\r\nTrace ID: ****\r\nCorrelation ID: *****\r\nTimestamp: 2022-12-16 17:18:11Z\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```
decoded:
```
ClientSecretCredential.get_token failed: Authentication failed: AADSTS500011: The resource principal named https://management.azure.com was not found in the tenant named PVUECN. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.
Trace ID: ****
Correlation ID: ****
Timestamp: 2022-12-16 16:08:27Z
Traceback (most recent call last):
  File \"/root/.ansible/tmp/ansible-tmp-1671206898.300529-135-138589534907732/AnsiballZ_azure_rm_virtualmachine.py\", line 107, in <module>
    _ansiballz_main()
  File \"/root/.ansible/tmp/ansible-tmp-1671206898.300529-135-138589534907732/AnsiballZ_azure_rm_virtualmachine.py\", line 99, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File \"/root/.ansible/tmp/ansible-tmp-1671206898.300529-135-138589534907732/AnsiballZ_azure_rm_virtualmachine.py\", line 47, in invoke_module
    runpy.run_module(mod_name='ansible_collections.azure.azcollection.plugins.modules.azure_rm_virtualmachine', init_globals=dict(_module_fqn='ansible_collections.azure.azcollection.plugins.modules.azure_rm_virtualmachine', _modlib_path=modlib_path),
  File \"/usr/lib64/python3.8/runpy.py\", line 207, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File \"/usr/lib64/python3.8/runpy.py\", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File \"/usr/lib64/python3.8/runpy.py\", line 87, in _run_code
    exec(code, run_globals)
  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_jbwif2z3/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 2344, in <module>
  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_jbwif2z3/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 2340, in main
  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_jbwif2z3/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 963, in __init__
  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_jbwif2z3/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/module_utils/azure_rm_common.py\", line 469, in __init__
  File \"/tmp/ansible_azure.azcollection.azure_rm_virtualmachine_payload_jbwif2z3/ansible_azure.azcollection.azure_rm_virtualmachine_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_virtualmachine.py\", line 1114, in exec_module
  File \"/usr/local/lib/python3.8/site-packages/azure/core/tracing/decorator.py\", line 78, in wrapper_use_tracer
    return func(*args, **kwargs)
  File \"/usr/local/lib/python3.8/site-packages/azure/mgmt/compute/v2021_04_01/operations/_virtual_machines_operations.py\", line 1502, in get
    pipeline_response = self._client._pipeline.run(request, stream=False, **kwargs)
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 211, in run
    return first_node.send(pipeline_request)  # type: ignore
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 71, in send
    response = self.next.send(request)
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 71, in send
    response = self.next.send(request)
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/_base.py\", line 71, in send
    response = self.next.send(request)
  [Previous line repeated 2 more times]
  File \"/usr/local/lib/python3.8/site-packages/azure/mgmt/core/policies/_base.py\", line 47, in send
    response = self.next.send(request)
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_redirect.py\", line 158, in send
    response = self.next.send(request)
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_retry.py\", line 446, in send
    response = self.next.send(request)
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_authentication.py\", line 116, in send
    self.on_request(request)
  File \"/usr/local/lib/python3.8/site-packages/azure/core/pipeline/policies/_authentication.py\", line 93, in on_request
    self._token = self._credential.get_token(*self._scopes)
  File \"/usr/local/lib/python3.8/site-packages/azure/identity/_internal/get_token_mixin.py\", line 76, in get_token
    token = self._request_token(*scopes, **kwargs)
  File \"/usr/local/lib/python3.8/site-packages/azure/identity/_internal/decorators.py\", line 56, in wrapper
    return fn(*args, **kwargs)
  File \"/usr/local/lib/python3.8/site-packages/azure/identity/_internal/client_credential_base.py\", line 40, in _request_token
    raise ClientAuthenticationError(message=message)
azure.core.exceptions.ClientAuthenticationError: Authentication failed: AADSTS500011: The resource principal named https://management.azure.com was not found in the tenant named PVUECN. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant.
```

**Comment by dbanttari** [comment]
Additional info: auth is done via a service principal.  The following environment variables are set:
```bash
  docker run --rm -it -u 0 \
  -e AZURE_CLIENT_ID \
  -e AZURE_SECRET \
  -e AZURE_CLIENT_SECRET \
  -e AZURE_TENANT \
  -e AZURE_TENANT_ID \
  -e AZURE_CLOUD_ENVIRONMENT \
  -e AZURE_CLOUD_NAME \
  -e AZURE_SUBSCRIPTION_ID \
  [etc]
```
(Some vars are duplicates of each other, eg AZURE_SECRET == AZURE_CLIENT_SECRET and AZURE_CLOUD_NAME and AZURE_CLOUD_ENVIRONMENT are in this case both set to `AzureChinaCloud`)
Other things that interact with Azure successfully run, such as the inventory script and an azure_keyvault_secret lookup, and the first task in the following test also works fine:
```
---
- name: Test azurerm issue
  hosts: all
  tasks:
  - name: Ensure other azurerm modules (in china) work correctly ** succeeds **
    azure.azcollection.azure_rm_resourcegroup_info:
      name: "{{ resource_group }}"
      list_resources: true
    delegate_to: localhost
  - name: Tell Azure to reboot this VM ** FAILS **
    azure.azcollection.azure_rm_virtualmachine:
      resource_group: "{{ resource_group }}"
      name: "{{ name }}"
      restarted: true
      zones: "{{ availability_zone }}"
    delegate_to: localhost
    connection: local

```
(edit: added second task that fails)

**Comment by dbanttari** [comment]
Just in case this is a problem in `azure-mgmt-compute`, we have v26.1.0 installed.  Here's the line from the stack trace: https://github.com/Azure/azure-sdk-for-python/blob/azure-mgmt-compute_26.1.0/sdk/compute/azure-mgmt-compute/azure/mgmt/compute/v2021_04_01/operations/_virtual_machines_operations.py#L1502

**Comment by dbanttari** [comment]
`azure_rm_virtualmachine_info` fails in the same way, but this works:
```
  - name: Ensure other azurerm modules (in china) work correctly  ** succeeds **
    azure.azcollection.azure_rm_resource_info:
      provider: compute
      resource_group: "{{ resource_group }}"
      resource_type: virtualmachines
      resource_name: "{{ name }}"
      api_version: "2021-04-01"
    delegate_to: localhost
```

**Comment by dbanttari** [comment]
As a workaround, anyone who only wants to update tags (and perhaps other things) can use azure_rm_resource instead:
```yaml
- name: Set patch_version tags
  azure.azcollection.azure_rm_resource:
    provider: compute
    resource_group: "{{ resource_group }}"
    resource_type: virtualmachines
    resource_name: "{{ name }}"
    api_version: "2021-04-01"
    method: PATCH
    body:
      tags: "{{ tags | combine( {'patch_version': ansible_date_time.iso8601} ) }}"
  delegate_to: localhost
```

**Comment by dbanttari** [comment]
Rebooting can also be accomplished via azure_rm_resource:
```yaml
  - name: Reboot instance
    azure.azcollection.azure_rm_resource:
      provider: compute
      resource_group: "{{ resource_group }}"
      resource_type: virtualmachines
      resource_name: "{{ name }}/restart"
      api_version: "2022-08-01"
      method: POST
    delegate_to: localhost
```

**Comment by Fred-sun** [comment]
@dbanttari I tested locally and did not encounter your problem. Did I encounter this problem before upgrading to v1.14.0

**Comment by dbanttari** [comment]
@Fred-sun tested with 1.13 and I have the same problem

**Comment by Fred-sun** [comment]
@dbanttari Can you try upgrading it to the latest version? Thanks!
```e.g.
ansible-galaxy collection install azure.azcollection --force

**Comment by dbanttari** [comment]
I initially reported this against the latest version (1.14).  In any case, I refreshed the image we're using for this (based on awx-ee) and am now getting a different error:

> ClientAuthenticationError: Authentication failed: AADSTS70011: The provided request must include a 'scope' input parameter. The provided value for the input parameter 'scope' is not valid. The scope https://management.chinacloudapi.cn.default is not valid.

I believe the root cause of this new error was fixed here: https://github.com/ansible-collections/azure/pull/984#event-7890331039 (Hi, @ChandlerSwift!)

Looks like there hasn't been a new version of azcollection published since that was merged.  Are you expecting to drop 1.15 anytime soon?
